UI Customization document - initial chapters (#35579)
keycloak/keycloak: docs/documentation/ui_customization/topics/localization.adoc

[[_localization]]
== Internationalization and Localization

Keycloak supports localized text throughout all user interfaces.

.Prerequisites

> * To enable internationalization for a realm, you must enable internationalization in the `Realm settings` section of the admin console. See the {adminguide_link}[{adminguide_name}].

Reviewer: "You must" sounds controlling.  "You" alone means the same thing.

> * To enable internationalization for a realm, you enable internationalization in the `Realm settings` section of the Admin Console. See {adminguide_link}#enabling-internationalization[{Enabling internationalization}].

Reviewer: ```suggestion
* You enable internationalization for a realm in the `Realm settings` section of the Admin Console. See {adminguide_link}#enabling-internationalization[{Enabling internationalization}].
```

=== Localizing mesages in a theme

> Text in the templates is loaded from message bundles. A theme that extends another theme inherits all messages from the parent's message bundle. You can

Reviewer: ```suggestion
Text in the templates is loaded from message bundles. When a theme extends another theme, the child theme inherits all messages from the parent's message bundle. You can
```

override individual messages by adding `<THEME TYPE>/messages/messages_en.properties` to your theme.

For example, to replace `Username` on the login form with `Your Username` for the `mytheme`, create the file
`themes/mytheme/login/messages/messages_en.properties` with the following content:

[source]
----
usernameOrEmail=Your Username
----

Within a message, values like `{0}` and `{1}` are replaced with arguments when the message is used. For example {0} in `Log in to {0}` is replaced with the name
of the realm.

> Texts of these message bundles can be overwritten by realm-specific values. The realm-specific values are manageable by the UI and API.

Reviewer: ```suggestion
Texts of these message bundles can be overwritten by realm-specific values, which are manageable by the UI and API.
```

=== Adding a language to a theme

.Procedure

. Create the file `<THEME TYPE>/messages/messages_<LOCALE>.properties` in the directory of your theme.

> . Add this file to the `locales` property in `<THEME TYPE>/theme.properties`.
> For a language to be available to users the realms `login`, `account` and `email`, the theme has to support the language, so you need to add your language for those theme types.

Reviewer: ```suggestion
. Add this file to the `locales` property in `<THEME TYPE>/theme.properties`.
+
For a language to be available to users and the `login`, `account` and `email` realms, the theme has to support the language, so you need to add your language for those theme types.
```

Author: > Rewording is my best guess.

Those are not realms.  They are theme types.  But it's a little confusing because as I recall, they are just referred to as `themes`.

This language comes from the current documentation, so I think it's safe to leave it alone for now.

Reviewer: Okay, not sure what it should be, but it seems there is missing punctuation somewhere.  
I was thinking, it should be "For a language to be available to users COMMA..." but I not sure.

> For a language to be available to users the realms `login`, `account` and `email`, the theme has to support the language, so you need to add your language for those theme types.

Reviewer: ```suggestion
For a language to be available to users in a realm, the login, account, and email theme types must support the language, so you need to add your language for those theme types.
```

Reviewer: On more attempt.

+

> For example, to add Norwegian translations to the `mytheme` theme create the file `themes/mytheme/login/messages/messages_no.properties` with the

Reviewer: ```suggestion
For example, to add Norwegian translations to the `mytheme` theme, create the file `themes/mytheme/login/messages/messages_no.properties` with the
```

following content:
+
[source]
----
usernameOrEmail=Brukernavn
password=Passord
----
+

> If you omit a translation for messages, they will use English.

Reviewer: ```suggestion
If you omit a translation for messages, those messages appear in English.
```

. Edit `themes/mytheme/login/theme.properties` and add:
+
[source]
----
locales=en,no
----

> . Add the same for the `account` and `email` theme types. To do this create `themes/mytheme/account/messages/messages_no.properties` and

Reviewer: ```suggestion
. Make the same changes to the `account` and `email` theme types. To make these changes, create `themes/mytheme/account/messages/messages_no.properties` and
```

> `themes/mytheme/email/messages/messages_no.properties`. Leaving these files empty will result in the English messages being used.

Reviewer: ```suggestion
`themes/mytheme/email/messages/messages_no.properties`. If you leave these files empty, English messages are used.
```

. Copy `themes/mytheme/login/theme.properties` to `themes/mytheme/account/theme.properties` and `themes/mytheme/email/theme.properties`.

> . Add a translation for the language selector. This is done by adding a message to the English translation. To do this add the following to
> `themes/mytheme/account/messages/messages_en.properties` and `themes/mytheme/login/messages/messages_en.properties`:

Reviewer: ```suggestion
. Add a translation for the language selector by adding a message to the English translation. Add the following to
`themes/mytheme/account/messages/messages_en.properties` and `themes/mytheme/login/messages/messages_en.properties`:
```

+
[source]
----
locale_no=Norsk
----

By default, message properties files should be encoded using UTF-8.
Keycloak falls back to ISO-8859-1 handling if it can't read the contents as UTF-8.
Unicode characters can be escaped as described in Java's documentation for https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/PropertyResourceBundle.html[PropertyResourceBundle].

> Previous versions of Keycloak supported specifying the encoding in the first line with a comment like `# encoding: UTF-8`, which is no longer supported.

Reviewer: ```suggestion
Previous versions of {project_name} supported specifying the encoding in the first line with a comment such as `# encoding: UTF-8`, which is no longer supported.
```

[role="_additional-resources"]
.Additional resources

> * See <<_locale_selector,Locale Selector>> for details on how the current locale is selected.

Reviewer: ```suggestion
* For details on how the current locale is selected, see <<_locale_selector,Locale Selector>>.
```

=== Overriding localized text for an entire realm

> Occasionally, you wish to declare translations that take effect for an entire realm.  This is done by specifying key/value pairs in the realm settings.  Any key specified in this way will override key/value pairs deployed with a theme.
> 

Reviewer: ```suggestion
You also have the option to declare translations that take effect for an entire realm.  You specify key/value pairs in the realm settings.  Any key specified in this way overrides the key/value pairs deployed with a theme.

```

> CAUTION: Using realm overrides is not the typical way to do localization in {project_name}.  Consider carefully if you want every theme in your realm to use these key/value pairs.

Reviewer: ```suggestion
CAUTION: Using realm overrides is not the recommended wayt to achieve localization in {project_name}.  Consider carefully if you want every theme in your realm to use key/value pairs.
```

.Procedure to add a realm override

> . Log into the Admin Console
> . Select your realm

Reviewer: ```suggestion
. Log into the Admin Console.
. Select your realm.
```

. Click *Realm Settings* from the menu.
. Click on the *Localization* tab.
. Click on the *Realm overrides* subtab.

> . Select a language from the dropdown
> . Click on *Add translation*
> . Create a key/value pair from the modal dialog

Reviewer: ```suggestion
. Select a language from the option list.
. Click *Add translation*.
. Create a key/value pair from the modal dialog.
```

> Note that there is another subtab called *Effective message bundles*.  This subtab provides a tool to query key/value pairs for a combination of theme, language, and theme type.  You can use this tool to test and make sure your realm overrides have taken effect.

Reviewer: ```suggestion
Notice another subtab called *Effective message bundles*.  This subtab provides a tool to query key/value pairs for a combination of theme, language, and theme type.  You can use this tool to test and make sure your realm overrides took effect.
```
